src/URI.js

```diff
@@ -1024,11 +1034,41 @@
       if (!punycode) {
         throw new TypeError('Hostname "' + v + '" contains characters other than [A-Z0-9.-] and Punycode.js is not available');
       }
-
       if (punycode.toASCII(v).match(URI.invalid_hostname_characters)) {
-        throw new TypeError('Hostname "' + v + '" contains characters other than [A-Z0-9.-]');
+        throw new TypeError('Hostname "' + v + '" contains characters other than [A-Z0-9.:-]');
+      }
+    }
+  };
+
+  URI.ensureValidPort = function(v) {
+    var valid = false;
+
+    if (!v || !v.length) {
+      // no custom port specified
+      valid = true;
+    } else {
+      var port = Number(v);
+
+      // verify type and range
+      if (Number.isInteger(port) && (port > 0) && (port < 65536)
+      ) {
+        valid = true;
       }
     }
+
+    if (!valid) {
+      throw new TypeError('Port "' + v + '" is not a valid port');
+    }
+  };
+
+  URI.basicValidation = function(parts) {
```

Comment:
this is not limited to HTTP/S, but also applies to ssh, ftp, and many more. Maybe we should create `URI.hostProtocols` analoguous to `URI.defaultPorts`? The name of this function should be similar to the property, so `URI.ensureHostProtocol()` or something like that? either way, `basic` is not a thing in URLs, so the term is really irritating.